File: python/y2017/d01/aoc_2017_01.py
import sys
from common import timer, get_input


def clean_input(data):
    return data[0]


@timer(part=1)
def solve_part_1(data, step=1):
    """Solution for part 1"""

    return sum(
        [
            int(c1)
            for c1, c2 in zip(data, f"{data[step:]}{data[:step]}")
            if c1 == c2
        ]
    )


@timer(part=2)
def solve_part_2(data):
    """Solution for part 2"""
    return solve_part_1(data, len(data) // 2)


@timer(part='main', title='Total elapsed', show_return=False)
def main(filename):
    data = clean_input(get_input(filename, 2017, 1))

    part1 = solve_part_1(data)
    part2 = solve_part_2(data)

    return part1, part2


if __name__ == "__main__":
    filename = "input.txt"
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    main(filename)

File: C++/2017/01/aoc_2017_01.cpp
#include "aoc_io.hpp"
#include "aoc_timer.hpp"
#include <algorithm>
#include <string>

namespace aoc_2017_01 {

int solution(std::string line, int step) {
    int sum = 0;

    std::string line2 = line;
    std::rotate(line2.begin(), line2.begin() + step, line2.end());
    for (std::string::iterator it1 = line.begin(), it2 = line2.begin(); it1 != line.end(); it1++, it2++) {
        if (*it1 == *it2) {
            sum += *it1 - '0';
        }
    }

    return sum;
}

int solve_part_1(const std::string &line) {
    return solution(line, 1);
}

int solve_part_2(const std::string &line) {
    return solution(line, static_cast<int>(line.length()) / 2);
}

int solve_all(const std::string &line) {
    aoc::timer(1, aoc_2017_01::solve_part_1, line);
    aoc::timer(2, aoc_2017_01::solve_part_2, line);

    return 0;
}
} // namespace aoc_2017_01

int main(int argc, char **argv) {
    std::string filename;

    const int year = 2017;
    const int day  = 1;

    std::vector<std::string> data;

    if (argc > 1) {
        filename = argv[1];
    } else {
        filename = "input.txt";
    }

    data = aoc::io::get_input_list<std::string>(filename, year, day);

    aoc::io::header(year, day);
    aoc::timer(0, aoc_2017_01::solve_all, data[0], false);

    return 0;
}
